In [1]:
import pandas as pd


In [2]:
df_customers = pd.read_csv("../data_raw/tonic_fabricate_db/customer.csv")

In [3]:
df_customers.head()

customer_id customer_number       customer_name creation_date   department  \
0            1         CL00001  Los Pinos Hermanos    2012-12-27   Montevide0   
1            2         CL00002     Don Juan & Cía.    2021-03-10    Río Negro   
2            3         CL00003     San José & Cía.    2014-04-26      Soriano   
3            4         CL00004    Los Pinos & Cía.    2010-09-21  Cerro Largo   
4            5         CL00005    El Ombú Hermanos    2024-06-15       Rivera   

            city         address           phone  \
0     Montevideo    Ruta 2 Km 34  +598 80 605971   
1          Young   Ruta 6 Km 193  +598 31 428763   
2        Cardona  Ruta 11 Km 107  +598 17 584083   
3  Fraile Muerto    Ruta 11 Km 1  +598 37 373306   
4     Tranqueras  Ruta 65 Km 170  +598 47 425856   

                              email  status  
0  CONTACTO@LOSPINOSHERMANOS.COM.UY  Activo  
1         contacto@donjuanca.com.uy  Activo  
2          contacto@sanjosca.com.uy  Activo  
3        contacto@lospinosca.com.uy  Activo  
4     contacto@elombhermanos.com.uy  Activo

In [4]:
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80 entries, 0 to 79
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   customer_id      80 non-null     int64 
 1   customer_number  80 non-null     object
 2   customer_name    80 non-null     object
 3   creation_date    80 non-null     object
 4   department       80 non-null     object
 5   city             80 non-null     object
 6   address          77 non-null     object
 7   phone            75 non-null     object
 8   email            78 non-null     object
 9   status           80 non-null     object
dtypes: int64(1), object(9)
memory usage: 6.4+ KB


In [5]:
df_customers.describe(include="all").T

count unique                       top freq  mean      std  \
customer_id      80.0    NaN                       NaN  NaN  40.5  23.2379   
customer_number    80     80                   CL00001    1   NaN      NaN   
customer_name      80     51           San José & Cía.    4   NaN      NaN   
creation_date      80     80                2012-12-27    1   NaN      NaN   
department         80     15                Tacuarembó   11   NaN      NaN   
city               80     31                Montevideo   10   NaN      NaN   
address            77     76             Ruta 17 Km 60    2   NaN      NaN   
phone              75     75            +598 80 605971    1   NaN      NaN   
email              78     53  contacto@sanjosca.com.uy    3   NaN      NaN   
status             80      2                    Activo   76   NaN      NaN   

                 min    25%   50%    75%   max  
customer_id      1.0  20.75  40.5  60.25  80.0  
customer_number  NaN    NaN   NaN    NaN   NaN  
customer_name    NaN    NaN   NaN    NaN   NaN  
creation_date    NaN    NaN   NaN    NaN   NaN  
department       NaN    NaN   NaN    NaN   NaN  
city             NaN    NaN   NaN    NaN   NaN  
address          NaN    NaN   NaN    NaN   NaN  
phone            NaN    NaN   NaN    NaN   NaN  
email            NaN    NaN   NaN    NaN   NaN  
status           NaN    NaN   NaN    NaN   NaN

In [6]:
df_customers.isna().sum().sort_values(ascending=False)


phone              5
address            3
email              2
customer_id        0
customer_number    0
customer_name      0
city               0
department         0
creation_date      0
status             0
dtype: int64

In [7]:
df_customers.duplicated().sum()

np.int64(0)

## Observaciones – Tabla Customers

- La tabla contiene 80 registros y 10 columnas.
- No se detectaron duplicados exactos en los registros.
- Se identificaron valores nulos en campos no críticos:
  - phone (5 registros)
  - address (3 registros)
  - email (2 registros)
- Las claves `customer_id` y `customer_number` no presentan valores nulos.

Se detecta repetición de valores en `customer_name`, lo que indica que este campo no puede utilizarse como identificador único.  
Para relaciones entre tablas se recomienda utilizar `customer_number`.


In [8]:
df_customers.groupby("customer_name")["customer_number"].nunique().sort_values(ascending=False).head(10)


customer_name
San José & Cía.      4
Celulosa & Cía.      3
El Ombú Hermanos     3
San José Ltda.       3
Maderas S.R.L.       3
San José S.R.L.      3
Santa María Ltda.    3
Forestal S.R.L.      3
Bosques S.A.         3
El Ombú Ltda.        2
Name: customer_number, dtype: int64

In [9]:
df_customers[df_customers["customer_name"] == "San José & Cía."]


customer_id customer_number    customer_name creation_date  department  \
2             3         CL00003  San José & Cía.    2014-04-26     Soriano   
37           38         CL00038  San José & Cía.    2014-10-24   Río Negro   
41           42         CL00042  San José & Cía.    2022-04-09    Paysandú   
50           51         CL00051  San José & Cía.    2021-08-28  Tacuarembó   

          city         address           phone                     email  \
2      Cardona  Ruta 11 Km 107  +598 17 584083  contacto@sanjosca.com.uy   
37       Young  Ruta 31 Km 113  +598 47 238525  contacto@sanjosca.com.uy   
41   Quebracho   Ruta 17 Km 67  +598 48 316237                       NaN   
50  Tacuarembó  Ruta 43 Km 177             NaN  contacto@sanjosca.com.uy   

    status  
2   Activo  
37  Activo  
41  Activo  
50  Activo

### Análisis

- Vemos que los datos tienen distinto ID, distinto customer_number, distintas fechas de creación, distintos departamentos ciudades y teléfonos, pero contacto por correo iguales. En un contexto real, este patrón podría significar que son sucursales de una misma firma, pero al tratarse de datos creados con IA y no estar seguros, tomé la decisión de tratarlos por customer_number como clientes independientes.


In [12]:
df_equipment = pd.read_csv("../data_raw/tonic_fabricate_db/equipment.csv")

In [13]:
df_equipment.head()


equipment_id serial_number          model  year  customer_id  \
0             1   PN201400001        Mammoth  2014            4   
1             2   PN201500002           Ergo  2015            4   
2             3   PN201500003  Elephant King  2015            4   
3             4   PN201700004       Scorpion  2017            4   
4             5   PN201400005       Scorpion  2014            4   

  current_status  hours_worked location_department          location_city  
0  En reparación         54614            Paysandú              Quebracho  
1  En reparación         50645               Salto                  Belén  
2         Activo         50792      Treinta y Tres  Santa Clara de Olimar  
3         Activo         39898         Cerro Largo                   Melo  
4         Activo         54039             Durazno                Durazno

In [14]:
df_equipment.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 570 entries, 0 to 569
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   equipment_id         570 non-null    int64 
 1   serial_number        570 non-null    object
 2   model                570 non-null    object
 3   year                 570 non-null    int64 
 4   customer_id          570 non-null    int64 
 5   current_status       570 non-null    object
 6   hours_worked         570 non-null    int64 
 7   location_department  570 non-null    object
 8   location_city        570 non-null    object
dtypes: int64(4), object(5)
memory usage: 40.2+ KB


In [15]:
df_equipment.describe(include="all").T


count unique          top freq          mean  \
equipment_id         570.0    NaN          NaN  NaN         285.5   
serial_number          570    570  PN201400001    1           NaN   
model                  570      5         Ergo  185           NaN   
year                 570.0    NaN          NaN  NaN   2017.438596   
customer_id          570.0    NaN          NaN  NaN     33.717544   
current_status         570      3       Activo  462           NaN   
hours_worked         570.0    NaN          NaN  NaN  37824.240351   
location_department    570     10  Cerro Largo   77           NaN   
location_city          570     30         Melo   31           NaN   

                              std     min      25%      50%      75%      max  
equipment_id           164.689101     1.0   143.25    285.5   427.75    570.0  
serial_number                 NaN     NaN      NaN      NaN      NaN      NaN  
model                         NaN     NaN      NaN      NaN      NaN      NaN  
year                     4.614935  2010.0   2013.0   2017.0   2022.0   2025.0  
customer_id              22.05065     1.0     14.0     30.0     55.0     80.0  
current_status                NaN     NaN      NaN      NaN      NaN      NaN  
hours_worked         23066.201028     0.0  15623.0  39427.5  59321.5  75934.0  
location_department           NaN     NaN      NaN      NaN      NaN      NaN  
location_city                 NaN     NaN      NaN      NaN      NaN      NaN

In [16]:
df_equipment["serial_number"].duplicated().sum()


np.int64(0)

In [18]:
df_equipment["customer_id"].isna().sum()


np.int64(0)

### Conclusiones – Equipment

- La tabla contiene 570 equipos sin valores nulos ni duplicados.
- Equipment_id actúa como clave técnica y serial_number como identificador lógico único.
- La flota se compone de 5 modelos, siendo Ergo el más frecuente.
- La antigüedad promedio es del año 2017, con equipos entre 2010 y 2025.
- Todos los equipos están asociados a un cliente (customer_id), lo que asegura integridad referencial.
- La mayoría de los equipos se encuentran en estado Activo.
- Las horas trabajadas muestran una amplia dispersión, habilitando análisis de desgaste y mantenimiento.

In [19]:
df_failures = pd.read_csv("../data_raw/tonic_fabricate_db/failures.csv")

In [20]:
df_failures.head()

failure_id  equipment_id        date failure_category severity  \
0           1           169  2021-06-08           Sensor     Alta   
1           2           254  2021-02-22        Eléctrico     Alta   
2           3           124  2024-12-26            Motor    Media   
3           4           316  2020-07-05       Hidráulico    Media   
4           5           440  2020-03-27            Motor     Baja   

        description  
0  Falla registrada  
1  Falla registrada  
2  Falla registrada  
3  Falla registrada  
4  Falla registrada

In [22]:
df_failures.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40 entries, 0 to 39
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   failure_id        40 non-null     int64 
 1   equipment_id      40 non-null     int64 
 2   date              40 non-null     object
 3   failure_category  40 non-null     object
 4   severity          40 non-null     object
 5   description       40 non-null     object
dtypes: int64(2), object(4)
memory usage: 2.0+ KB


In [23]:
df_failures.describe(include="all").T

count unique               top freq     mean         std  \
failure_id        40.0    NaN               NaN  NaN     20.5   11.690452   
equipment_id      40.0    NaN               NaN  NaN  282.325  160.966534   
date                40     40        2021-06-08    1      NaN         NaN   
failure_category    40      4             Motor   14      NaN         NaN   
severity            40      3              Alta   15      NaN         NaN   
description         40      1  Falla registrada   40      NaN         NaN   

                   min     25%    50%    75%    max  
failure_id         1.0   10.75   20.5  30.25   40.0  
equipment_id      12.0  136.75  302.5  422.5  569.0  
date               NaN     NaN    NaN    NaN    NaN  
failure_category   NaN     NaN    NaN    NaN    NaN  
severity           NaN     NaN    NaN    NaN    NaN  
description        NaN     NaN    NaN    NaN    NaN

### Conclusiones – Failures

- La tabla contiene 40 registros de fallas asociadas a equipos.
- No se detectan valores nulos ni filas duplicadas.
- La columna "date" se encuentra como texto y deberá ser convertida a fecha en etapas posteriores.
- Existen 4 categorías de falla, siendo “Motor” la más frecuente.
- La severidad “Alta” es la más común.
- La columna "description" no aporta variabilidad (un único valor).


In [24]:
df_parts_sales = pd.read_csv("../data_raw/tonic_fabricate_db/parts_sales.csv")


In [25]:
df_parts_sales.head()

sale_line_id   sale_date  customer_id  part_id  quantity   unit_price  \
0             1  2020-10-02           55     5852         3  2511.734613   
1             2  2020-10-02           55     4234         1  2273.598479   
2             3  2020-06-27            4     2051         1  2345.632810   
3             4  2020-06-27            4     5676         2  2848.529555   
4             5  2024-09-20           63     4705         1   645.215897   

   total_price sale_type  service_id invoice_number  
0  7535.203839  Servicio         1.0       INV-1000  
1  2273.598479  Servicio         1.0       INV-1001  
2  2345.632810  Servicio         2.0       INV-1002  
3  5697.059109  Servicio         2.0       INV-1003  
4   645.215897  Servicio         3.0       INV-1004

In [26]:
df_parts_sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8620 entries, 0 to 8619
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   sale_line_id    8620 non-null   int64  
 1   sale_date       8620 non-null   object 
 2   customer_id     8620 non-null   int64  
 3   part_id         8620 non-null   int64  
 4   quantity        8620 non-null   int64  
 5   unit_price      8620 non-null   float64
 6   total_price     8620 non-null   float64
 7   sale_type       8620 non-null   object 
 8   service_id      6120 non-null   float64
 9   invoice_number  8620 non-null   object 
dtypes: float64(3), int64(4), object(3)
memory usage: 673.6+ KB


In [27]:
df_parts_sales.describe(include="all").T

count unique         top  freq         mean          std  \
sale_line_id    8620.0    NaN         NaN   NaN       4310.5  2488.523994   
sale_date         8620   1849  2023-10-27    18          NaN          NaN   
customer_id     8620.0    NaN         NaN   NaN    35.446172    22.853478   
part_id         8620.0    NaN         NaN   NaN   3013.07297  1739.604383   
quantity        8620.0    NaN         NaN   NaN     2.305684     1.131638   
unit_price      8620.0    NaN         NaN   NaN   1845.05753  1358.690212   
total_price     8620.0    NaN         NaN   NaN  4340.361087  4330.902466   
sale_type         8620      2    Servicio  6120          NaN          NaN   
service_id      6120.0    NaN         NaN   NaN  1526.956863   864.584576   
invoice_number    8620   8620    INV-1000     1          NaN          NaN   

                       min          25%          50%          75%  \
sale_line_id           1.0      2155.75       4310.5      6465.25   
sale_date              NaN          NaN          NaN          NaN   
customer_id            1.0         14.0         32.0         55.0   
part_id                2.0      1497.75       3017.0      4505.25   
quantity               1.0          1.0          2.0          3.0   
unit_price      -69.825536   833.110356  1545.660585  2556.187412   
total_price    -333.423511  1441.098982  2957.582668   5754.56446   
sale_type              NaN          NaN          NaN          NaN   
service_id             1.0       780.75       1539.0       2276.0   
invoice_number         NaN          NaN          NaN          NaN   

                         max  
sale_line_id          8620.0  
sale_date                NaN  
customer_id             80.0  
part_id               6000.0  
quantity                 5.0  
unit_price       7582.915504  
total_price     37914.577522  
sale_type                NaN  
service_id            3000.0  
invoice_number           NaN

### Observaciones – Sales / Sale Lines

- La tabla contiene 8.620 líneas de venta de repuestos.
- No se detectan valores nulos excepto en "service_id", lo cual es consistente con ventas no asociadas a servicios.
- "sale_date" se encuentra como texto y deberá convertirse a fecha.
- Se detectan valores negativos en "unit_price" y "total_price", posiblemente asociados a devoluciones o ajustes.
- Existen dos tipos de venta, predominando las ventas asociadas a servicios.

In [28]:
df_parts_usage = pd.read_csv("../data_raw/tonic_fabricate_db/parts_usage.csv")

In [29]:
df_parts_usage.head()

part_usage_id  service_id  part_id  quantity   unit_price  total_price  \
0              1           1     5852       3.0  2511.734613  7535.203839   
1              2           1     4234       1.0  2273.598479  2273.598479   
2              3           2     2051       1.0  2345.632810  2345.632810   
3              4           2     5676       2.0  2848.529555  5697.059109   
4              5           3     4705       1.0   645.215897   645.215897   

  warehouse_location  
0                 A1  
1                 A1  
2                 A1  
3                 A1  
4                 A1

In [30]:
df_parts_usage.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7549 entries, 0 to 7548
Data columns (total 7 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   part_usage_id       7549 non-null   int64  
 1   service_id          7549 non-null   int64  
 2   part_id             7549 non-null   int64  
 3   quantity            7298 non-null   float64
 4   unit_price          7549 non-null   float64
 5   total_price         7549 non-null   float64
 6   warehouse_location  7300 non-null   object 
dtypes: float64(3), int64(3), object(1)
memory usage: 413.0+ KB


In [31]:
df_parts_usage.describe(include="all").T

count unique  top  freq         mean          std  \
part_usage_id       7549.0    NaN  NaN   NaN       3775.0  2179.352924   
service_id          7549.0    NaN  NaN   NaN  1516.999868   861.946905   
part_id             7549.0    NaN  NaN   NaN  3010.312889  1735.052146   
quantity            7298.0    NaN  NaN   NaN     2.004522        0.815   
unit_price          7549.0    NaN  NaN   NaN  1702.579884  1267.724629   
total_price         7549.0    NaN  NaN   NaN  3397.407783  3042.938464   
warehouse_location    7300      1   A1  7300          NaN          NaN   

                           min          25%          50%          75%  \
part_usage_id              1.0       1888.0       3775.0       5662.0   
service_id                 1.0        774.0       1523.0       2260.0   
part_id                    2.0       1499.0       3027.0       4482.0   
quantity                   1.0          1.0          2.0          3.0   
unit_price          -64.927395   765.729473   1426.84242  2346.022802   
total_price        -194.064616  1266.077533  2525.334004  4683.058614   
warehouse_location         NaN          NaN          NaN          NaN   

                             max  
part_usage_id             7549.0  
service_id                3000.0  
part_id                   6000.0  
quantity                     3.0  
unit_price           6896.678688  
total_price         20618.046112  
warehouse_location           NaN

### Observaciones – Part Usage

- La tabla contiene 7.549 registros de uso de repuestos dentro de servicios.
- Todos los registros están asociados a un "service_id", lo cual es consistente con el proceso operativo.
- Existen valores nulos en "quantity" y "warehouse_location", lo que sugiere registros incompletos o campos poco utilizados.
- Se detectan valores negativos en "unit_price" y "total_price", consistentes con ajustes o devoluciones internas.
- La ubicación de almacén presenta un único valor dominante, por lo que su valor analítico es limitado.


In [33]:
df_parts = pd.read_csv("../data_raw/tonic_fabricate_db/parts.csv")

In [34]:
df_parts.head()

part_id part_number                    description     category  unit_cost  \
0        1   PN-000001    Piñón Transmisión - Premium  Transmisión     604.82   
1        2   PN-000002    Módulo Eléctrica - Standard    Electrica     240.17   
2        3   PN-000003      Inyector Motor - Original        Motor    2577.45   
3        4   PN-000004       Bujía Motor - Compatible        Motor    1013.14   
4        5   PN-000005  Sensor Eléctrica - Compatible    Eléctrica    1479.88   

               supplier  stock_qty  min_stock  
0  Ponsse Latin America        103         10  
1  Ponsse Latin America         72         20  
2              Tecnifor          3          8  
3             Ponsse Oy         44         18  
4          Nordic Parts          9         23

In [35]:
df_parts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6000 entries, 0 to 5999
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   part_id      6000 non-null   int64  
 1   part_number  6000 non-null   object 
 2   description  6000 non-null   object 
 3   category     6000 non-null   object 
 4   unit_cost    6000 non-null   float64
 5   supplier     5823 non-null   object 
 6   stock_qty    6000 non-null   int64  
 7   min_stock    6000 non-null   int64  
dtypes: float64(1), int64(3), object(4)
memory usage: 375.1+ KB


In [36]:
df_parts.describe(include="all").T

count unique                          top  freq         mean  \
part_id      6000.0    NaN                          NaN   NaN       3000.5   
part_number    6000   6000                    PN-000001     1          NaN   
description    6000    160  Batería Eléctrica - Premium    57          NaN   
category       6000      7                   Hidráulica  1542          NaN   
unit_cost    6000.0    NaN                          NaN   NaN  1742.428527   
supplier       5823      6                    Ponsse Oy  1026          NaN   
stock_qty    6000.0    NaN                          NaN   NaN    54.713667   
min_stock    6000.0    NaN                          NaN   NaN      14.6155   

                     std   min       25%      50%        75%      max  
part_id      1732.195139   1.0   1500.75   3000.5    4500.25   6000.0  
part_number          NaN   NaN       NaN      NaN        NaN      NaN  
description          NaN   NaN       NaN      NaN        NaN      NaN  
category             NaN   NaN       NaN      NaN        NaN      NaN  
unit_cost    1241.128916 -50.0  807.1475  1449.76  2423.8275  5498.27  
supplier             NaN   NaN       NaN      NaN        NaN      NaN  
stock_qty      34.034204   0.0      25.0     54.0       84.0    123.0  
min_stock       5.780966   5.0      10.0     15.0       20.0     24.0

### Observaciones - Parts

- La tabla "parts" contiene 6000 registros y no presenta valores nulos en "part_id" ni en "part_number", por lo que ambos campos funcionan correctamente como identificadores.
- Se detecta una alta repetición en la columna "description", lo que indica variantes de repuestos o una generación de datos no totalmente normalizada.
- La columna "unit_cost" presenta valores negativos, lo que representa una inconsistencia que deberá corregirse en la etapa de limpieza.
- Existen valores nulos en la columna "supplier", lo que sugiere información incompleta que deberá evaluarse posteriormente.
- Los campos de inventario ("stock_qty" y "min_stock") son coherentes en general y permiten identificar posibles situaciones de stock crítico.

In [37]:
df_sales = pd.read_csv("../data_raw/tonic_fabricate_db/sales.csv")

In [38]:
df_sales.head()

sale_id  customer_id     model      serial_number  year   sale_date  \
0        1           53    Wisent  PN2022WISSALE0001  2022  2023-02-02   
1        2           71   Buffalo  PN2024BUFSALE0002  2024  2022-04-24   
2        3           66  Scorpion  PN2025SCOSALE0003  2025  2021-11-23   
3        4           62      Bear  PN2022BEASALE0004  2022  2020-05-21   
4        5           12   Mammoth  PN2025MAMSALE0005  2025  2020-02-01   

   sale_price  warranty_months     payment_terms delivery_date  \
0    534103.0               36           Contado    2023-03-07   
1    481632.0               24           90 días    2022-06-05   
2    432812.0               36           Contado    2021-12-25   
3    486352.0               24           Contado    2020-07-15   
4    598472.0               24  Leasing 36 meses    2020-04-30   

       salesperson  
0   Eduardo Castro  
1  Patricia Suárez  
2  Roberto Mendoza  
3  Roberto Mendoza  
4              NaN

In [39]:
df_sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 202 entries, 0 to 201
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   sale_id          202 non-null    int64  
 1   customer_id      202 non-null    int64  
 2   model            202 non-null    object 
 3   serial_number    202 non-null    object 
 4   year             202 non-null    int64  
 5   sale_date        202 non-null    object 
 6   sale_price       202 non-null    float64
 7   warranty_months  202 non-null    int64  
 8   payment_terms    202 non-null    object 
 9   delivery_date    202 non-null    object 
 10  salesperson      198 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 17.5+ KB


In [40]:
df_sales.describe(include="all").T

count unique                top freq           mean  \
sale_id          202.0    NaN                NaN  NaN          101.5   
customer_id      202.0    NaN                NaN  NaN      40.782178   
model              202      7      Elephant King   36            NaN   
serial_number      202    202  PN2022WISSALE0001    1            NaN   
year             202.0    NaN                NaN  NaN    2022.549505   
sale_date          202    191         2024-05-24    2            NaN   
sale_price       202.0    NaN                NaN  NaN  512038.717822   
warranty_months  202.0    NaN                NaN  NaN      22.811881   
payment_terms      202      7            60 días   36            NaN   
delivery_date      202    189         2024-06-25    2            NaN   
salesperson        198      5         Laura Vega   47            NaN   

                          std       min        25%       50%        75%  \
sale_id             58.456537       1.0      51.25     101.5     151.75   
customer_id         21.877817       1.0      22.25      43.5      60.75   
model                     NaN       NaN        NaN       NaN        NaN   
serial_number             NaN       NaN        NaN       NaN        NaN   
year                 1.828472    2020.0     2021.0    2023.0     2024.0   
sale_date                 NaN       NaN        NaN       NaN        NaN   
sale_price       54776.995854  407426.0  471674.25  501964.5  554688.25   
warranty_months      9.943835      12.0       12.0      24.0       36.0   
payment_terms             NaN       NaN        NaN       NaN        NaN   
delivery_date             NaN       NaN        NaN       NaN        NaN   
salesperson               NaN       NaN        NaN       NaN        NaN   

                      max  
sale_id             202.0  
customer_id          76.0  
model                 NaN  
serial_number         NaN  
year               2025.0  
sale_date             NaN  
sale_price       658878.0  
warranty_months      36.0  
payment_terms         NaN  
delivery_date         NaN  
salesperson           NaN

### Observaciones - Sales

- La tabla "service_order" contiene 202 registros y no presenta valores nulos en "sale_id" ni en "serial_number", por lo que los identificadores son consistentes.
- La columna "model" presenta una cantidad limitada de valores únicos, lo que indica un catálogo de equipos bien definido.
- Existen valores nulos en la columna "salesperson", lo que representa información comercial incompleta que deberá tratarse en la etapa de limpieza.
- Las columnas de fechas ("sale_date" y "delivery_date") están almacenadas como texto y deberán convertirse a tipo fecha.
- Los valores de "sale_price" y "warranty_months" son coherentes en general y permiten análisis comerciales y de garantía.

In [41]:
df_service_order = pd.read_csv("../data_raw/tonic_fabricate_db/service_order.csv")

In [42]:
df_service_order.head()

service_id  equipment_id  customer_id service_type  start_date    end_date  \
0           1           130           55   Correctivo  2020-10-02  2020-10-05   
1           2            23            4   Preventivo  2020-06-27         NaN   
2           3            36           63   Preventivo  2024-09-20  2024-09-30   
3           4            12            4   Preventivo  2025-05-10         NaN   
4           5           296           29   Correctivo  2024-10-21         NaN   

                  reported_issue       root_cause      technician      status  
0     Necesita reemplazo por uso  Desgaste Normal  Diego Martínez     Cerrado  
1  Mantenimiento programado 500h              NaN  Fernando Silva  En Proceso  
2     Cambio de aceite y filtros  Desgaste Normal      Juan Pérez     Cerrado  
3  Mantenimiento programado 500h              NaN  Fernando Silva     Abierto  
4      Falla en panel de control              NaN      Juan Pérez  En Proceso

In [43]:
df_service_order.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   service_id      3000 non-null   int64 
 1   equipment_id    3000 non-null   int64 
 2   customer_id     3000 non-null   int64 
 3   service_type    3000 non-null   object
 4   start_date      3000 non-null   object
 5   end_date        2117 non-null   object
 6   reported_issue  3000 non-null   object
 7   root_cause      2124 non-null   object
 8   technician      2950 non-null   object
 9   status          3000 non-null   object
dtypes: int64(3), object(7)
memory usage: 234.5+ KB


In [44]:
df_service_order.describe(include="all").T

count unique                            top  freq  \
service_id      3000.0    NaN                            NaN   NaN   
equipment_id    3000.0    NaN                            NaN   NaN   
customer_id     3000.0    NaN                            NaN   NaN   
service_type      3000      3                     Preventivo  1229   
start_date        3000   1559                     2022-07-27     7   
end_date          2117   1341                     2021-04-22     5   
reported_issue    3000     35  Inspección general preventiva   219   
root_cause        2124      7                Desgaste Normal  1039   
technician        2950      5                 Fernando Silva   608   
status            3000      3                        Cerrado  2156   

                      mean         std  min     25%     50%      75%     max  
service_id          1500.5  866.169729  1.0  750.75  1500.5  2250.25  3000.0  
equipment_id    286.149333  163.533781  1.0   143.0   292.0    424.0   570.0  
customer_id      33.718667    22.32031  1.0    14.0    29.0     55.0    80.0  
service_type           NaN         NaN  NaN     NaN     NaN      NaN     NaN  
start_date             NaN         NaN  NaN     NaN     NaN      NaN     NaN  
end_date               NaN         NaN  NaN     NaN     NaN      NaN     NaN  
reported_issue         NaN         NaN  NaN     NaN     NaN      NaN     NaN  
root_cause             NaN         NaN  NaN     NaN     NaN      NaN     NaN  
technician             NaN         NaN  NaN     NaN     NaN      NaN     NaN  
status                 NaN         NaN  NaN     NaN     NaN      NaN     NaN

### Observaciones – Service Orders

- La tabla contiene 3.000 registros de órdenes de servicio.
- Todos los registros están asociados a un equipment_id y un customer_id, lo que permite relacionarlos con equipos y clientes.
- El tipo de servicio más frecuente es "Preventivo", seguido por "Correctivo".
- Existen valores nulos en end_date, lo que indica órdenes aún abiertas o en proceso.
- El campo root_cause presenta valores faltantes, principalmente en servicios preventivos.
- Se observa una alta concentración del estado "Cerrado", lo que sugiere buen cierre operativo de órdenes.
- Hay una cantidad menor de valores nulos en technician, lo que podría deberse a registros incompletos.

In [45]:
df_warranty_cases = pd.read_csv("../data_raw/tonic_fabricate_db/warranty_cases.csv")

In [46]:
df_warranty_cases.head()

warranty_id  equipment_id  customer_id case_open_date case_close_date  \
0            1           331           70     2025-01-13      2025-02-02   
1            2           262           10     2024-05-07             NaN   
2            3           149           45     2023-04-25      2023-05-03   
3            4           542           51     2022-03-17      2022-03-23   
4            5           388           22     2025-12-26      2026-01-03   

     issue_description approval_status  amount_claimed  amount_approved  
0  Reclamo de garantía        Aprobado            4746      3760.801866  
1  Reclamo de garantía       Pendiente            5155         0.000000  
2  Reclamo de garantía       Rechazado            1443         0.000000  
3  Reclamo de garantía       Rechazado            1308         0.000000  
4  Reclamo de garantía        Aprobado            4408      3712.377110

In [47]:
df_warranty_cases.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   warranty_id        350 non-null    int64  
 1   equipment_id       350 non-null    int64  
 2   customer_id        350 non-null    int64  
 3   case_open_date     350 non-null    object 
 4   case_close_date    266 non-null    object 
 5   issue_description  350 non-null    object 
 6   approval_status    350 non-null    object 
 7   amount_claimed     350 non-null    int64  
 8   amount_approved    350 non-null    float64
dtypes: float64(1), int64(4), object(4)
memory usage: 24.7+ KB


In [48]:
df_warranty_cases.describe(include="all").T

count unique                  top freq         mean  \
warranty_id        350.0    NaN                  NaN  NaN        175.5   
equipment_id       350.0    NaN                  NaN  NaN   284.685714   
customer_id        350.0    NaN                  NaN  NaN    35.374286   
case_open_date       350    321           2020-05-04    2          NaN   
case_close_date      266    253           2020-06-18    3          NaN   
issue_description    350      1  Reclamo de garantía  350          NaN   
approval_status      350      3             Aprobado  178          NaN   
amount_claimed     350.0    NaN                  NaN  NaN      2941.82   
amount_approved    350.0    NaN                  NaN  NaN  1241.392491   

                           std    min      25%         50%          75%  \
warranty_id         101.180532    1.0    88.25       175.5       262.75   
equipment_id        165.673406    1.0    146.0       282.0       434.75   
customer_id          22.148494    1.0     15.0        34.5         55.0   
case_open_date             NaN    NaN      NaN         NaN          NaN   
case_close_date            NaN    NaN      NaN         NaN          NaN   
issue_description          NaN    NaN      NaN         NaN          NaN   
approval_status            NaN    NaN      NaN         NaN          NaN   
amount_claimed     1461.033302  506.0  1661.25      2986.5      4132.25   
amount_approved    1499.539805    0.0      0.0  527.355504  2481.151552   

                           max  
warranty_id              350.0  
equipment_id             570.0  
customer_id               80.0  
case_open_date             NaN  
case_close_date            NaN  
issue_description          NaN  
approval_status            NaN  
amount_claimed          5496.0  
amount_approved    5199.786159

### Observaciones – Warranty Cases

- La tabla contiene 350 registros de casos de garantía.
- Todos los casos están asociados a un equipment_id y un customer_id, lo que permite su análisis cruzado con equipos y clientes.
- Existen valores nulos en case_close_date, lo que indica casos de garantía aún abiertos o en estado pendiente.
- El estado de aprobación más frecuente es "Aprobado", seguido por "Rechazado" y "Pendiente".
- Se observa que muchos registros tienen amount_approved igual a 0, consistente con casos rechazados o aún no resueltos.
- El monto aprobado suele ser inferior al monto reclamado, lo que sugiere ajustes parciales en las garantías.
- El campo issue_description presenta un único valor, por lo que su aporte analítico es limitado.